src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -141,13 +164,57 @@ static void splitMutationCommands(
             {
                 if (command.type == MutationCommand::Type::READ_COLUMN)
                     for_interpreter.push_back(command);
-                else if (command.type == MutationCommand::Type::RENAME_COLUMN)
-                    part_columns.rename(command.column_name, command.rename_to);
 
                 for_file_renames.push_back(command);
             }
         }
+
+        auto alter_conversions = part->storage.getAlterConversionsForPart(part);
+        /// We don't add renames from commands, instead we take them from rename_map.
+        /// It's important because required renames depend not only on part's data version (i.e. mutation version)
+        /// but also on part's metadata version. Why we have such logic only for renames? Because all other types of alter
+        /// can be deduced based on difference between part's schema and table schema.
+
+        for (const auto & [rename_to, rename_from] : alter_conversions.rename_map)
+        {
+            for_file_renames.push_back({.type = MutationCommand::Type::RENAME_COLUMN, .column_name = rename_from, .rename_to = rename_to});
+        }
+    }
+}
+
+/// It's legal to squash renames because commands with rename are always "barrier"
+/// and executed separately from other types of commands.
+static MutationCommands squashRenamesInCommands(const MutationCommands & commands)
```

Comment:
Does it really make sense? As I understand from your [code](https://github.com/ClickHouse/ClickHouse/blob/1683cd5d6fedbf5b388f0d5275062b37ea0c3c10/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1848-L1861) renames cannot be squashed together even with renames from another alter and transitive renames in one alter are not [allowed](https://github.com/ClickHouse/ClickHouse/blob/0f182f61641adadcbdd028c4e31b2ae3b5a24fbf/src/Storages/AlterCommands.cpp#L1217-L1228) https://github.com/ClickHouse/ClickHouse/blob/1683cd5d6fedbf5b388f0d5275062b37ea0c3c10/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1848-L1861.